crates/ibc/src/clients/ics07_tendermint/client_state.rs

```diff
@@ -253,6 +254,83 @@ impl ClientState {
             _ => Ok(()),
         }
     }
+
+    fn check_trusted_header(
+        consensus_state: &TmConsensusState,
+        header: &Header,
+    ) -> Result<(), Ics02Error> {
+        let trusted_val_hash = header.trusted_validator_set.hash();
+
+        if consensus_state.next_validators_hash != trusted_val_hash {
+            return Err(Error::misbehaviour_trusted_validator_hash_mismatch(
+                header.trusted_validator_set.clone(),
+                consensus_state.next_validators_hash,
+                trusted_val_hash,
+            )
+            .into());
+        }
+
+        Ok(())
+    }
+
+    fn check_misbehaviour_header(
+        &self,
+        header: &Header,
+        consensus_state: &TmConsensusState,
+        current_timestamp: Timestamp,
+    ) -> Result<(), Ics02Error> {
+        Self::check_trusted_header(consensus_state, header)?;
+
+        let duration_since_consensus_state = current_timestamp
+            .duration_since(&consensus_state.timestamp())
+            .ok_or_else(|| {
+                Ics02Error::invalid_consensus_state_timestamp(
+                    consensus_state.timestamp(),
+                    current_timestamp,
+                )
+            })?;
+
+        if duration_since_consensus_state >= self.trusting_period {
+            return Err(
+                Error::misbehaviour_consensus_state_timestamp_gte_trusting_period(
```

Comment:
...and we can return `invalid_consensus_state_timestamp` or some other error that doesn't mention `misbehaviour`

Reply:
https://github.com/cosmos/ibc-rs/pull/215/commits/396c07f77d0a00ebe6b489d5186e548c9d57a43f